clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -604,6 +605,11 @@ private void readBufferedCompletedFetchesForPausedTopicPartitions(TopicPartition
 
                         try {
                             nextInLineRecords = parseCompletedFetch(completedFetch);
+
+                            // nextInLineRecords might be null when completedFetch contains error
+                            if (nextInLineRecords == null) {
+                                completedFetch.response.decRefCount();
```

Comment:
Isn't it cleaner to increment the count in the PartitionRecords constructor and decrement here across all scenarios?

Reply:
Alright, moved outside of the block

Reply:
It is slightly tricky. I thought about it a bit more. The reason that won't work is because then NetworkThread callback would decrement the count to zero for ByteBuffer after calling future.complete and return to the MemoryPool. However, the buffer is still in use.